File: src/main/scala/structural/facade/ShopFacade.scala
package structural.facade

class ShopFacade
(
   paymentService: PaymentService = new PaymentService(),
   shippingService: ShippingService = new ShippingService(),
   smsService: SmsService = new SmsService(),
) {

  def placeOrderOnline(email: String, address: String): Unit = {
    println("Bắt đầu quy trình mua hàng trực tuyến...")
    paymentService.paymentByOnlineBanking()
    shippingService.shipping(address)
    smsService.sendSMS(email)
    println("Bạn đã nhận được hàng!")
  }

  def placeOrderOffline(): Unit = {
    println("Bắt đầu quy trình mua hàng trực tiếp...")
    paymentService.paymentByCash()
    println("Bạn đã nhận được hàng!")
  }
}

File: src/structural/facade/ShopFacade.java
package structural.facade;

public class ShopFacade {
    PaymentService paymentService;
    ShippingService shippingService;
    SmsService smsService;

    public ShopFacade() {
        this.paymentService = new PaymentService();
        this.shippingService = new ShippingService();
        this.smsService = new SmsService();
    }

    public void placeOrderOnline(String email, String address) {
        System.out.println("Placing order online...");
        paymentService.paymentByOnlineBanking();
        shippingService.shipping(address);
        smsService.sendSMS(email);
        System.out.println("Package received!");
    }

    public void placeOrderOffline() {
        System.out.println("Placing order offline...");
        paymentService.paymentByOnlineBanking();
        System.out.println("Package received!");
    }
}
